src/VisualStudio/Core/Def/Implementation/TableDataSource/Suppression/VisualStudioSuppressionFixService.cs

```diff
@@ -117,50 +122,109 @@ public bool RemoveSuppressions(bool selectedErrorListEntriesOnly, IVsHierarchy p
+            Action<CancellationToken> computeDiagnosticsToFix = cancellationToken =>
+            {
+                // If we are fixing selected diagnostics in error list, then get the diagnostics from error list entry snapshots.
+                // Otherwise, get all diagnostics from the diagnostic service.
+                var diagnosticsToFixTask = selectedEntriesOnly ?
+                    _suppressionStateService.GetSelectedItemsAsync(isAddSuppression, cancellationToken) :
+                    GetAllDiagnosticsAsync(shouldFixInProject, cancellationToken);
+
+                diagnosticsToFix = diagnosticsToFixTask.WaitAndGetResult(cancellationToken);
+            };
+
+            var title = GetFixTitle(isAddSuppression);
+            var waitDialogMessage = GetWaitDialogMessage(isAddSuppression);
+            var result = InvokeWithWaitDialog(computeDiagnosticsToFix, title, waitDialogMessage);
+
+            // Bail out if the user cancelled.
+            if (result == WaitIndicatorResult.Canceled)
+            {
+                return null;
+            }
+
+            return diagnosticsToFix;
+        }
+
         private bool ApplySuppressionFix(Func<Project, bool> shouldFixInProject, bool selectedEntriesOnly, bool isAddSuppression, bool isSuppressionInSource, bool onlyCompilerDiagnostics, bool showPreviewChangesDialog)
         {
+            var diagnosticsToFix = GetDiagnosticsToFix(shouldFixInProject, selectedEntriesOnly, isAddSuppression);
+            return ApplySuppressionFix(diagnosticsToFix, shouldFixInProject, selectedEntriesOnly, isSuppressionInSource, isAddSuppression, onlyCompilerDiagnostics, showPreviewChangesDialog);
+        }
+
+        private bool ApplySuppressionFix(IEnumerable<DiagnosticData> diagnosticsToFix, Func<Project, bool> shouldFixInProject, bool filterStaleDiagnostics, bool isAddSuppression, bool isSuppressionInSource, bool onlyCompilerDiagnostics, bool showPreviewChangesDialog)
+        {
+            if (diagnosticsToFix == null)
+            {
+                return false;
+            }
+
+            diagnosticsToFix = FilterDiagnostics(diagnosticsToFix, isAddSuppression, isSuppressionInSource, onlyCompilerDiagnostics);
+            if (diagnosticsToFix.IsEmpty())
+            {
+                // Nothing to fix.
+                return true;
+            }
+
             ImmutableDictionary<Document, ImmutableArray<Diagnostic>> documentDiagnosticsToFixMap = null;
             ImmutableDictionary<Project, ImmutableArray<Diagnostic>> projectDiagnosticsToFixMap = null;
 
-            var title = isAddSuppression ? ServicesVSResources.SuppressMultipleOccurrences : ServicesVSResources.RemoveSuppressMultipleOccurrences;
-            var waitDialogMessage = isAddSuppression ? ServicesVSResources.ComputingSuppressionFix : ServicesVSResources.ComputingRemoveSuppressionFix;
+            var title = GetFixTitle(isAddSuppression);
```

Comment:
Basically, we use diagnostic data until the point where we are building up the data structure needed by something that needs diagnostic. In this case, the documentDiagnosticsToFixMap and projectDiagnosticsToFixMap are needed by Workspace layer FixMultipleContext (flavor of FixAllContext with specified set of diagnostics), so we use diagnostic.